lib/improver/tests/nbhood/circular_kernel/test_circular_kernel.py

```diff
@@ -0,0 +1,183 @@
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Unit tests for the nbhood.circular_kernel.circular_kernel function."""
+
+import unittest
+
+from iris.tests import IrisTest
+import numpy as np
+
+from improver.nbhood.circular_kernel import circular_kernel
+
+
+class Test_circular_kernel(IrisTest):
+
+    """Test neighbourhood processing plugin."""
+
+    def test_basic(self):
+        """Test that the plugin returns a Numpy array."""
+        ranges = (2, 2)
+        fullranges = (0, 2, 2)
+        weighted_mode = False
+        result = (
+            circular_kernel(fullranges, ranges, weighted_mode))
+        self.assertIsInstance(result, np.ndarray)
+
+    def test_single_point_weighted(self):
+        """Test behaviour for a unitary range, with weighting."""
+        ranges = (1, 1)
+        fullranges = (1, 1)
+        weighted_mode = True
+        expected = [[0., 0., 0.], [0., 1., 0.], [0., 0., 0.]]
+        result = (
+            circular_kernel(fullranges, ranges, weighted_mode))
+        self.assertArrayAlmostEqual(result, expected)
+
+    def test_single_point_unweighted(self):
+        """Test behaviour for a unitary range without weighting.
+        Note that this gives one more grid cell range than weighted! As the
+        affected area is one grid cell more in each direction, an equivalent
+        range of 2 was chosen for this test."""
+        ranges = (1, 1)
+        fullranges = (1, 1)
+        weighted_mode = False
+        expected = [[0., 1., 0.], [1., 1., 1.], [0., 1., 0.]]
+        result = (
+            circular_kernel(fullranges, ranges, weighted_mode))
+        self.assertArrayEqual(result, expected)
+
+    def test_range5_weighted(self):
+        """Test behaviour for a range of 5, with weighting."""
+        ranges = (5, 5)
+        fullranges = (5, 5)
+        weighted_mode = True
+        expected = [
+            [0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.],
+            [0.,   0.,   0.,   0.2,  0.32, 0.36, 0.32, 0.2,  0.,   0.,   0.],
+            [0.,   0.,   0.28, 0.48, 0.6,  0.64, 0.6,  0.48, 0.28, 0.,   0.],
+            [0.,   0.2,  0.48, 0.68, 0.8,  0.84, 0.8,  0.68, 0.48, 0.2,  0.],
```

Comment:
Plyint complains because of lots of wrong spacing, (although the wrong spacing is making it easier to read). It gives a score of -9  at the moment :(

Reply:
I haven't changed these. We probably need to confirm whether this formatting is acceptable.